integration/build_dependencies_test.go

```diff
@@ -96,9 +93,96 @@ func TestBuild_WithDependencies(t *testing.T) {
 	}
 }
 
+func TestBuildDependenciesCache(t *testing.T) {
+	MarkIntegrationTest(t, CanRunWithoutGcp)
+
+	// These tests build 4 images and then make a file change to the images in `change`.
+	// The test then triggers another build and verifies that the images in `rebuilt` were built
+	// (e.g., the changed images and their dependents), and that the other images were found in the artifact cache.
+	// It runs the profile `concurrency-0` which builds with maximum concurrency.
+	tests := []struct {
+		description string
+		change      []int
+		rebuilt     []int
+	}{
+		{
+			description: "no change",
+		},
+		{
+			description: "change 1",
+			change:      []int{1},
+			rebuilt:     []int{1},
+		},
+		{
+			description: "change 2",
+			change:      []int{2},
+			rebuilt:     []int{1, 2},
+		},
+		{
+			description: "change 3",
+			change:      []int{3},
+			rebuilt:     []int{1, 2, 3},
+		},
+		{
+			description: "change 4",
+			change:      []int{4},
+			rebuilt:     []int{4},
+		},
+		{
+			description: "change all",
+			change:      []int{1, 2, 3, 4},
+			rebuilt:     []int{1, 2, 3, 4},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.description, func(t *testing.T) {
+			skaffold.Build("--cache-artifacts=true", "-p", "concurrency-0").InDir("testdata/build-dependencies").RunOrFail(t)
+			checkImagesExist(t)
+
+			// modify file `foo` to invalidate cache for target artifacts
+			for _, i := range test.change {
+				Run(t, fmt.Sprintf("testdata/build-dependencies/app%d", i), "sh", "-c", fmt.Sprintf("echo %s > foo", uuid.New().String()))
+			}
+			out, err := skaffold.Build("--cache-artifacts=true", "-p", "concurrency-0").InDir("testdata/build-dependencies").RunWithCombinedOutput(t)
+			if err != nil {
+				t.Fatal("expected build to succeed")
+			}
+			log := string(out)
+
+			for i := 1; i <= 4; i++ {
+				if !contains(test.rebuilt, i) && !strings.Contains(log, fmt.Sprintf("image%d: Found Locally", i)) {
```

Comment:
+1 to not depend on output and consume the event api to make test if not a huge re-work.

Reply:
Any specific reason to not depend on log output? There a few tests that do that already.

Reply:
The tests become brittle.  Our log output isn't API.  Those other tests likely predate the event API, and should be changed.

Reply:
I was gonna start working with this idea but there's no event api for cache hits and misses.

Reply:
we'd have to code against `BuildStarted` present and absent(polling over a time window). I think that's a bit brittle too.

Reply:
I created issue https://github.com/GoogleContainerTools/skaffold/issues/4927 to add an event for cache hits. I'll add a TODO to change this test after that gets done.

Reply:
You *won't* get buildEvents for cached images.  So accumulate the buildEvents seen and then check that only the images expected were rebuilt.  And the order should be safe too when they have dependencies.

Reply:
+1, that's similar to what's happening in the `rpc_test.go` integration test as well. we shouldn't need to add an event for cache hits/misses unless we think we'll use it outside of tests.

Reply:
I get the sentiment of this request but hear me out:
I'm testing 6 scenarios. It's necessary to test all of them. If I read from the log output then I can immediately determine if the cache was hit or miss and return immediately. The overall test takes ~30 seconds. 
If I swap this out with an rpc client then:
- I cannot test the `all artifacts found in cache` scenario, since that build completes so quickly then we can't even establish an rpc client connection in time before the build is complete.
- I need to artificially run every sub test for a fixed amount of time with a sufficient buffer so as to ***not make it flaky*** (which is the point of your argument against the log approach). With 5 scenarios and a 30 sec window per build thats 2.5 minutes.
- "Log output isn't an API contract". Agreed but that doesn't make the test flaky. If someone changes the log output then they'll also have to change the test accordingly. There is no scenario where the test might intermittently fail under the current approach.

Some of these points could be alleviated with an explicit event for build cache hits but then again point 1 can't be fixed and I don't want to skip a test case to adhere to _best practice_.

cc @tstromberg